### DISCUSSION TITLE
Ability to setSearchParams without navigation in useSearchParams

### DISCUSSION THREAD

**Discussion by KevinVandy** [remix-run/remix#7966] [Proposals]
It appears that the current intended functionality of the `useSearchParams` `setSearchParams` function is to always cause a navigation and trigger loaders whenever `setSearchParams` is called. I think it would be useful to be able to prevent a navigation and loaders being triggered. Perhaps a `navigate` or `navigation` boolean option.

For example, here is what I was hoping I could write:
```ts
const [searchParams, setSearchParams] = useSearchParams();

const currentStep = searchParams.get("currentStep");

const setCurrentStep = (step) => {
  setSearchParams((prev) => ({ ...prev, currentStep: step }), {
    navigation: false, //proposed new option to simply change url search params without navigation
    replace: true,
  });
};
```
 
I am, of course, able to achieve this today by following a pattern like this. The main reason for using search params state was to grab the initial state, but I don't need any of the route or root loaders to reload whenever this value changes in the URL.

```ts
  const [searchParams] = useSearchParams();

  const initialCurrentStep = searchParams.get("currentStep");

  const [currentStep, _setCurrentStep] = useState(initialCurrentStep);

  const setCurrentStep = (step: null | string) => {
    _setCurrentStep(step);
    history.replaceState(null, "", `?currentStep=${step}`);
  };
  ```

Maybe this isn't the intended use case for `useSearchParams`, but I think programmatically being able to control whether or not a navigation is needed upon a URL state change would be convenient.

**Comment by Somebi** [comment]
Completely agree. Example would be infinite scrolling with current page update in url. So that when user refresh a page via F5 or share a link, actual page parameter would be in url. Sending any server-side rendering requests not required for that.

**Comment by danthegoodman1** [comment]
this is a pain, how is this not solved?

**Comment by brophdawg11** [comment]
This discussion is being closed as part of the Remix v3 repository cleanup.

Remix v1/v2 was [upstreamed into React Router v7](https://remix.run/blog/wake-up-remix), so if this proposal is still relevant, please reopen or continue the discussion in the [React Router](https://github.com/remix-run/react-router) repository.
